diffusion/evaluation/clean_fid_eval.py

```diff
@@ -0,0 +1,237 @@
+                load_weights_only=True,
+                eval_dataloader=self.eval_dataloader)
+
+        # Move CLIP metric to device
+        self.device = dist.get_local_rank()
+        self.clip_metric = self.clip_metric.to(self.device)
+
+        # Predownload the CLIP model for computing clip-fid
+        _, _ = clip.load('ViT-B/32', device=self.device)
+
+    def _generate_images(self, guidance_scale: float):
+        """Core image generation function. Generates images at a given guidance scale.
+
+        Args:
+            guidance_scale (float): The guidance scale to use for image generation.
+        """
+        # Verify output dirs exist, if they don't, create them
+        real_image_path = os.path.join(self.output_dir, f'real_images_gs_{guidance_scale}')
+        gen_image_path = os.path.join(self.output_dir, f'gen_images_gs_{guidance_scale}')
+        if not os.path.exists(real_image_path) and dist.get_local_rank() == 0:
+            os.makedirs(real_image_path)
+        if not os.path.exists(gen_image_path) and dist.get_local_rank() == 0:
+            os.makedirs(gen_image_path)
+
+        # Reset the CLIP metric
+        self.clip_metric.reset()
+
+        # Storage for prompts
+        prompts = {}
+        # Iterate over the eval dataloader
+        num_batches = len(self.eval_dataloader)
+        for batch_id, batch in tqdm(enumerate(self.eval_dataloader)):
+            # Break if enough samples have been generated
+            if batch_id * self.batch_size * dist.get_world_size() >= self.num_samples:
+                break
+
+            real_images = batch['image']
+            captions = batch['captions']
+            # Ensure a new seed for each batch, as randomness in model.generate is fixed.
+            starting_seed = self.seed + num_batches * dist.get_local_rank()
+            seed = starting_seed + batch_id
+            # Generate images from the captions
+            with get_precision_context(self.precision):
+                generated_images = self.model.generate(tokenized_prompts=captions,
+                                                       height=self.size,
+                                                       width=self.size,
+                                                       guidance_scale=guidance_scale,
+                                                       seed=seed,
+                                                       progress_bar=False)  # type: ignore
+            # Get the prompts from the tokens
+            text_captions = [self.tokenizer.decode(caption, skip_special_tokens=True) for caption in captions]
+            self.clip_metric.update((generated_images * 255).to(torch.uint8), text_captions)
+            # Save the real images
+            for i, img in enumerate(real_images):
+                # Real images are -1, 1, so rescale to 0, 1.
+                to_pil_image((img + 1) / 2).save(f'{real_image_path}/{batch_id}_{i}_rank_{dist.get_local_rank()}.png')
```

Comment:
The real images will only be [-1, 1] if the images are normalized, do we want to make this assumption? It seems more natural to assume unnormalized images which can be done by specifying the appropriate transforms in `build_streaming_image_caption_dataloader()`

Reply:
We'll have to make an assumption here, or allow it to be configurable. Unnormalized images could mean multiple things depending on the specific dataset in question. Likely they are either in [0, 255] or [0, 1]. Should it be configurable? If not, what do you think the assumption should be?

Reply:
I'd rather make an assumption than make it configurable. 

The current assumption is that the images will be [-1, 1] due to normalization, right? I think [0, 1] would be better assumption. What do you think?

Reply:
Ok, switched to default of [0, 1] and raise a warning if images are not in this range.